feat: improved error responses

lidel opened the pull request:
This PR:


-  removes reliance on HTTP semantics for error codes and switches to a generic `HTTP 400` with `Error` object in body that is returned with mandatory `reason` string and optional `details`, as proposed in https://github.com/ipfs/pinning-services-api-spec/issues/57

- Every endpoint now has only two response types: _success_ and _error_, and success has the name of returned object in its description, making the spec docs easier to read:

![2020-09-02--12-54-07](https://user-images.githubusercontent.com/157609/91972873-9b158480-ed1b-11ea-9a59-4400b0ebc1ad.png)


**PREVIEW:** https://ipfs.github.io/pinning-services-api-spec/#specUrl=https://raw.githubusercontent.com/ipfs/pinning-services-api-spec/feat/error-codes/ipfs-pinning-service.yaml


Closes #57 @obo20 @andrew  @GregTheGreek @priom @jsign  @sanderpick @andrewxhill @ipfs/wg-pinning-services 

Please provide feedback (even if its just :+1: / :-1:)

obo20 commented:
This looks good to me!

andrew commented:
Only thing that springs to mind here is that sometimes the http infrastructure hosting the pinning service API may kick out a different 4XX or 5XX error code before the application server running the API gets chance to respond, clients should probably make sure they handle a variety of regular http errors along side the errors returned from IPFS itself.

jacobheun commented:
I think the change to the error response is good, but I don't think we should be getting rid of the other http status codes. At the very least we should retain the 500 response. Having each of the 400 codes we've already added helps to clarify the category of error that should be expected when implementing/using this api.

obo20 commented:
@lidel I agree with @jacobheun on this one. I think it would be beneficial to maintain the original 400-500 error codes that we had before.

priom commented:
This is a good and simple to start, but also  an application should handle a variety of 200, 400 & 500 for specific errors ans responses so client can be very informative about it.

lidel commented:
Thank you for feedback! 

- I've restored old error codes and added example for each:
  > ![2020-09-07--13-29-04](https://user-images.githubusercontent.com/157609/92383580-e14d5800-f10e-11ea-9765-cccf1beda0ad.png)

- I changed `500` to `5XX` to set proper expectations: reverse proxies may return all sorts of errors (503, 504 etc)


Q: is generic `400` enough as "catch-all" type of error? Or should we add separate `4XX` for "custom errors" not covered by the spec?

jacobheun commented:
> Q: is generic 400 enough as "catch-all" type of error? Or should we add separate 4XX for "custom errors" not covered by the spec?

I think adding the additional 4XX for other errors would be useful. To me this implies that while we've documented common errors you might see, the API may still return other errors, such as 405 which we haven't covered. Should these also be required to return the same error object format as the other, defined 4xx codes? I'm leaning yes, but am somewhat concerned about adding an early restriction on this for code ranges.

lidel commented:
Added `4XX:

![Screenshot_2020-09-07 IPFS Pinning Service API](https://user-images.githubusercontent.com/157609/92391156-70ad3800-f11c-11ea-8e19-2c92fd70f3e0.png)


As for suggesting specific response format for `4XX`, those should be ok, as this type of error will be returned by the pinning service app.

The problematic one may be `5XX`, where response can come from a reverse proxy that timeouts while trying to talk to the service app. One can configure proxy to return errors in specific format, but realistically one can't assume `5XX` error body will known, so clients usually check only the status code, body would be optional/best-effort.

andrew reviewed (approved):
👍
